### DISCUSSION TITLE
Easier debuggability for pending promises after a test has finished

### DISCUSSION THREAD

**Discussion by jespertheend** [denoland/deno#17808] [Ideas]
Sometimes in tests I run across the following message:
```
Promise resolution is still pending but the event loop has already resolved.
```
but I just can't pin down where the pending promise is being created.
Just now I simply forgot to call [`FakeTime.prototype.restore()`](https://deno.land/std@0.177.0/testing/time.ts?s=FakeTime#method_restore_0), but it took me a good couple of minutes to figure out what was happening.

I feel like Deno has some information here that could be communicated to the user to make debugging these cases easier. Like showing a stack trace to the pending promises.

Not sure if V8 exposes that sort of thing though, maybe what I'm asking here is impossible.

**Comment by bartlomieju** [comment]
This should be easy to fix - the error is returned in `JsRuntime::poll_value` - we should add a call to `find_stalled_top_level_await()` there that will give a stack trace for the promise that is unresolved. @jespertheend would you be interested in implementing that?

**Reply by jespertheend** [reply]
I mean, I can try 😅

**Reply by bartlomieju** [reply]
Just open a PR and we can work from there :)

**Reply by jespertheend** [reply]
Alright, I'll have some free time in a couple of hours.

**Reply by jespertheend** [reply]
I've submitted #17813 but am running into a panic, that I'm not really sure how to resolve: https://github.com/denoland/deno/pull/17813#issuecomment-1435246567
